tinynn/graph/quantization/utils.py

```diff
@@ -83,3 +83,66 @@ def clamp_with_fusion_(x: torch.Tensor, min_val: float, max_val: float) -> torch
     if not x.is_quantized:
         return torch.clamp_(x, min_val, max_val)
     return x
+
+
+def fake_quantize(tensor, asym, eps, quant_max, quant_min):
+    min_val, max_val = torch.aminmax(tensor)
+    device = tensor.device
+    zero_point = torch.zeros(min_val.size(), dtype=torch.int64, device=device)
+    if not asym:
+        max_val_pos = torch.max(-min_val, max_val)
+        scale = max_val_pos / (float(quant_max - quant_min) / 2)
+        scale = torch.max(scale, eps)
+    else:
+        scale = (max_val - min_val) / float(quant_max - quant_min)
+        scale = torch.max(scale, eps)
+        zero_point = quant_min - torch.round(min_val / scale).to(torch.int)
+        zero_point = torch.clamp(zero_point, quant_min, quant_max)
+    # do fake quantize
+    return torch.fake_quantize_per_tensor_affine(tensor, scale, zero_point, quant_min, quant_max)
+
+
+def lstm_weight_fake_quantize(model: torch.nn.LSTM, asym=False, bit=8, eps=torch.tensor([1e-6])):
+    '''Quantize the weights of LSTM individually to align with TFLite's implementation.'''
+
+    def _lstm_weight_fake_quantize(weight, asym=False, bit=8, eps=torch.tensor([1e-6])):
+        assert bit == 8
+        if weight.numel() == 0:
+            return weight
+        quant_min, quant_max = -127, 127
+        weight_shape = list(weight.size())
+        weight_parts = weight.split(weight_shape[0] // 4, dim=0)
+        weight_quant_parts = []
+        for i in range(4):
+            weight_quant_parts.append(
+                fake_quantize(weight_parts[i], asym=asym, eps=eps, quant_min=quant_min, quant_max=quant_max)
+            )
+        weight_quant = torch.cat(weight_quant_parts)
+        return weight_quant
+
+    for weight_name in ['weight_ih_l0', 'weight_hh_l0']:
+        quantized_weight = _lstm_weight_fake_quantize(getattr(model, weight_name), asym=asym, bit=bit, eps=eps)
+        setattr(model, f'{weight_name}_og', getattr(model, weight_name).clone())
```

Comment:
I think we should keep the modules clean. First of all, it's hacky. Second, it may lead to a name conflict if someone decided to use that name. Third, serialization/deserialization may fail. Maybe we should store it in the `Quantizer` instances.